### PR TITLE
functions: cloud bigtable reading function example

### PR DESCRIPTION
**Pull request by billyjacobson**

### CODE REVIEW

File: functions/bigtable/bigtable.go [hunk 1]
```diff
@@ -0,0 +1,65 @@
...
+// every request.
+var client *bigtable.Client
+var clientOnce sync.Once
+
+// BigtableRead is an example of reading Bigtable from a Cloud Function.
+func BigtableRead(w http.ResponseWriter, r *http.Request) {
+	clientOnce.Do(func() {
+		// Declare a separate err variable to avoid shadowing client.
+		var err error
+		client, err = bigtable.NewClient(r.Context(), r.Header.Get("projectId"), r.Header.Get("instanceId"))
```
Comment (tbpg) [inline functions/bigtable/bigtable.go:39]: I think this should be `context.Background()`?

File: functions/bigtable/bigtable.go [hunk 2]
```diff
@@ -0,0 +1,65 @@
...
+			osBuild := ""
+			for _, col := range row["stats_summary"] {
+				if col.Column == "stats_summary:os_build" {
+					osBuild = string(col.Value)
+				}
+			}
+
+			fmt.Fprintf(w, "Rowkey: %s, os_build:  %s\n", row.Key(), osBuild)
+			return true
+		}); err != nil {
```
Comment (tbpg) [inline functions/bigtable/bigtable.go:59]: Given the multi-line function, I think it makes sense to split this `err` declaration onto a separate line. Might help to keep the indentation clear. On first read, I forgot this was inside an `if` declaration.

File: functions/bigtable/bigtable_test.go [hunk 3]
```diff
@@ -0,0 +1,124 @@
...
+	"strings"
+	"testing"
+
+	"cloud.google.com/go/bigtable"
+	"github.com/google/uuid"
+)
+
+func TestBigtableRead(t *testing.T) {
+	ctx := context.Background()
+	project := os.Getenv("GOLANG_SAMPLES_BIGTABLE_PROJECT")
```
Comment (tbpg) [inline functions/bigtable/bigtable_test.go:32]: Call `testutil.SystemTest()`.

File: functions/bigtable/bigtable_test.go [hunk 4]
```diff
@@ -0,0 +1,124 @@
...
+	instance := os.Getenv("GOLANG_SAMPLES_BIGTABLE_INSTANCE")
+	if project == "" || instance == "" {
+		t.Skip("Skipping functions bigtable integration test. Set GOLANG_SAMPLES_BIGTABLE_PROJECT and GOLANG_SAMPLES_BIGTABLE_INSTANCE.")
+	}
+
+	adminClient, err := bigtable.NewAdminClient(ctx, project, instance)
+
+	uuid, err := uuid.NewRandom()
+	tableId := fmt.Sprintf("mobile-time-series-%s", uuid.String()[:8])
```
Comment (tbpg) [inline functions/bigtable/bigtable_test.go:41]:
```suggestion
	tableID := fmt.Sprintf("mobile-time-series-%s", uuid.String()[:8])
```

https://github.com/golang/go/wiki/CodeReviewComments#initialisms

Comment (tbpg) [inline functions/bigtable/bigtable_test.go:41]: Why `mobile-time-series`?

File: functions/bigtable/bigtable_test.go [hunk 5]
```diff
@@ -0,0 +1,124 @@
...
+	req := httptest.NewRequest("GET", "/", nil)
+	req.Header.Set("projectId", project)
+	req.Header.Set("instanceId", instance)
+	req.Header.Set("tableId", tableId)
+	rr := httptest.NewRecorder()
+	BigtableRead(rr, req)
+
+	t.Log(rr.Body.String())
+	got := (rr.Body.String())
+	if want :=
```
Comment (tbpg) [inline functions/bigtable/bigtable_test.go:65]: Would probably swap `got` inline and `want` outside to make it easier to follow the multi-line declaration.

File: functions/bigtable/bigtable_test.go [hunk 6]
```diff
@@ -0,0 +1,124 @@
...
+
+	req := httptest.NewRequest("GET", "/", nil)
+	req.Header.Set("projectId", project)
+	req.Header.Set("instanceId", instance)
+	req.Header.Set("tableId", tableId)
+	rr := httptest.NewRecorder()
+	BigtableRead(rr, req)
+
+	t.Log(rr.Body.String())
+	got := (rr.Body.String())
```
Comment (tbpg) [inline functions/bigtable/bigtable_test.go:64]:
```suggestion
	got := rr.Body.String()
```

File: functions/bigtable/bigtable_test.go [hunk 7]
```diff
@@ -0,0 +1,124 @@
...
+	writeTestData(err, ctx, project, instance, tableId, timestamp, t)
+
+	req := httptest.NewRequest("GET", "/", nil)
+	req.Header.Set("projectId", project)
+	req.Header.Set("instanceId", instance)
+	req.Header.Set("tableId", tableId)
+	rr := httptest.NewRecorder()
+	BigtableRead(rr, req)
+
+	t.Log(rr.Body.String())
```
Comment (tbpg) [inline functions/bigtable/bigtable_test.go:63]: Is this needed?

File: functions/bigtable/bigtable_test.go [hunk 8]
```diff
@@ -0,0 +1,124 @@
...
+	got := (rr.Body.String())
+	if want :=
+		`Rowkey: phone#4c410523#20190501, os_build:  PQ2A.190405.003
+Rowkey: phone#4c410523#20190502, os_build:  PQ2A.190405.004
+Rowkey: phone#4c410523#20190505, os_build:  PQ2A.190406.000
+Rowkey: phone#5c10102#20190501, os_build:  PQ2A.190401.002
+Rowkey: phone#5c10102#20190502, os_build:  PQ2A.190406.000`; !strings.Contains(got, want) {
+		t.Errorf("got %q, want %q", got, want)
```
Comment (tbpg) [inline functions/bigtable/bigtable_test.go:71]: Include the name of the function that failed in the error message.

File: functions/bigtable/bigtable_test.go [hunk 9]
```diff
@@ -0,0 +1,124 @@
...
+	}
+
+	if rr.Code != http.StatusOK {
+		t.Fatalf("BigtableRead got code %v, want %v", rr.Code, http.StatusOK)
+	}
+
+	adminClient.DeleteTable(ctx, tableId)
+}
+
+func writeTestData(err error, ctx context.Context, project string, instance string, tableId string, timestamp bigtable.Timestamp, t *testing.T) {
```
Comment (tbpg) [inline functions/bigtable/bigtable_test.go:81]: `ctx` should always be the first arg (Go convention).

**Comment by billyjacobson** [comment]
> Just to confirm, did you try deploying and using this?

yes I deployed and used it :)

File: functions/bigtable/bigtable_test.go (hunk 4 above)
Comment (billyjacobson) [reply functions/bigtable/bigtable_test.go:41]: that's just the example database we use for code samples

File: functions/bigtable/bigtable_test.go [hunk 10]
```diff
@@ -0,0 +1,124 @@
...
+
+func TestBigtableRead(t *testing.T) {
+	ctx := context.Background()
+	tc := testutil.SystemTest(t)
+	instance := os.Getenv("GOLANG_SAMPLES_BIGTABLE_INSTANCE")
+	if instance == "" {
+		t.Skip("Skipping functions bigtable integration test. Set GOLANG_SAMPLES_BIGTABLE_INSTANCE.")
+	}
+
+	adminClient, _ := bigtable.NewAdminClient(ctx, tc.ProjectID, instance)
```
Comment (tbpg) [inline functions/bigtable/bigtable_test.go:39]: Should probably check these errors -- I've seen them fail when the credentials don't have access to the project.

**Pull request merged**